tfx/experimental/pipeline_testing/stub_component_launcher.py

```diff
@@ -0,0 +1,99 @@
+
+import os
+from typing import Any, Dict, List, Text, Type
+
+from tfx import types
+from tfx.components.base import base_executor
+from tfx.experimental.pipeline_testing import base_stub_executor
+from tfx.orchestration.launcher import in_process_component_launcher
+
+class StubComponentLauncher(
+    in_process_component_launcher.InProcessComponentLauncher):
+  """Responsible for launching a stub executor.
+
+  The executor will be launched in the same process of the rest of the
+  component, i.e. its driver and publisher.
+  """
+
+  def _run_executor(self, execution_id: int,
+                    input_dict: Dict[Text, List[types.Artifact]],
+                    output_dict: Dict[Text, List[types.Artifact]],
+                    exec_properties: Dict[Text, Any]) -> None:
+    """Execute underlying component implementation."""
+    component_id = self._component_info.component_id
+    if component_id not in self.stubbed_component_map:
+      super(StubComponentLauncher, self)._run_executor(execution_id,
+                                                       input_dict,
+                                                       output_dict,
+                                                       exec_properties)
+    else:
+      executor_context = base_executor.BaseExecutor.Context(
+          beam_pipeline_args=self._beam_pipeline_args,
+          tmp_dir=os.path.join(self._pipeline_info.pipeline_root, '.temp', ''),
+          unique_id=str(execution_id))
+      executor = self.stubbed_component_map[component_id](component_id,
+                                                          self.test_data_dir,
+                                                          executor_context)
+      executor.Do(input_dict, output_dict, exec_properties)
+
+def get_stub_launcher_class(
+    test_data_dir: Text,
+    stubbed_component_ids: List[Text],
+    stubbed_component_map: Dict[Text,
+                                Type[base_stub_executor.BaseStubExecutor]]
+    ) -> Type[StubComponentLauncher]:
+  """Returns a StubComponentLauncher class.
+
+  Factory method returns a StubComponentLauncher class, which then can be
+  supported by the pipeline.
+  For example:
+    PipelineConfig(
+        supported_launcher_classes=[
+            stub_component_launcher.get_stub_launcher_class(
+                test_data_dir,
+                stubbed_component_ids,
+                stubbed_component_map)
+        ],
+    )
+  The method also sets the necessary class variables for the
+  StubComponentLauncher class, including stubbed_component_ids and
+  stubbed_component_map holding custom executor classes, which
+  users may define differently per component.
+
+  Args:
+    test_data_dir: The directory where pipeline outputs are recorded
+      (pipeline_recorder.py).
+    stubbed_component_ids: List of component ids that should be replaced
+      with a BaseStubExecutor.
+    stubbed_component_map: Dictionary holding user-defined stub executor.
```

Comment:
It is still unclear how the values in stubbed_component_map can be further customized. Maybe add an example in pydoc?